contracts/yellow-button/lib.rs

```diff
@@ -0,0 +1,545 @@
+    pub type Result<T> = core::result::Result<T, Error>;
+    /// Event type
+    type Event = <YellowButton as ContractEventBase>::Type;
+
+    impl From<InkEnvError> for Error {
+        fn from(e: InkEnvError) -> Self {
+            match e {
+                InkEnvError::Decode(_e) => {
+                    Error::ContractCall(String::from("Contract call failed due to Decode error"))
+                }
+                InkEnvError::CalleeTrapped => Error::ContractCall(String::from(
+                    "Contract call failed due to CalleeTrapped error",
+                )),
+                InkEnvError::CalleeReverted => Error::ContractCall(String::from(
+                    "Contract call failed due to CalleeReverted error",
+                )),
+                InkEnvError::KeyNotFound => Error::ContractCall(String::from(
+                    "Contract call failed due to KeyNotFound error",
+                )),
+                InkEnvError::_BelowSubsistenceThreshold => Error::ContractCall(String::from(
+                    "Contract call failed due to _BelowSubsistenceThreshold error",
+                )),
+                InkEnvError::TransferFailed => Error::ContractCall(String::from(
+                    "Contract call failed due to TransferFailed error",
+                )),
+                InkEnvError::_EndowmentTooLow => Error::ContractCall(String::from(
+                    "Contract call failed due to _EndowmentTooLow error",
+                )),
+                InkEnvError::CodeNotFound => Error::ContractCall(String::from(
+                    "Contract call failed due to CodeNotFound error",
+                )),
+                InkEnvError::NotCallable => Error::ContractCall(String::from(
+                    "Contract call failed due to NotCallable error",
+                )),
+                InkEnvError::Unknown => {
+                    Error::ContractCall(String::from("Contract call failed due to Unknown error"))
+                }
+                InkEnvError::LoggingDisabled => Error::ContractCall(String::from(
+                    "Contract call failed due to LoggingDisabled error",
+                )),
+                InkEnvError::EcdsaRecoveryFailed => Error::ContractCall(String::from(
+                    "Contract call failed due to EcdsaRecoveryFailed error",
+                )),
+                #[cfg(any(feature = "std", test, doc))]
+                InkEnvError::OffChain(_e) => {
+                    Error::ContractCall(String::from("Contract call failed due to OffChain error"))
+                }
+            }
+        }
+    }
+
+    /// Defines the storage
+    #[ink(storage)]
+    #[derive(SpreadAllocate)]
+    pub struct YellowButton {
+        /// access control
+        owner: AccountId,
+        /// How long does TheButton live for?
+        button_lifetime: u32,
+        /// is The Button dead
+        is_dead: bool,
+        /// block number at which the game ends
+        deadline: u32,
```

Comment:
This value seems to be redundant, it's always `last_press+deadline`. Not sure what the best practice here is: minimize number of storage items or storage queries?

Reply:
storage I suppose - the queries can be limited with some caching or building a view form the emitted events I suppose. Removed :shrug: